### PR TITLE
Synchronous search interface

### PR DESCRIPTION
**Pull request by PLNech**
- Exposed `Index.searchSync()` to search an Index synchronously, fixing #21.
- Raw methods have been exposed as protected for internal manipulation of request's raw bytes.

@clement-leprovost-algolia: Does this look good to you?

### CODE REVIEW

File: algoliasearch/src/main/java/com/algolia/search/saas/BaseAPIClient.java [hunk 1]
```diff
@@ -330,15 +349,75 @@ private String _getAnswer(InputStream istream) throws IOException {
...
+
+
     private JSONObject _getJSONObject(String input) throws JSONException {
         return new JSONObject(new JSONTokener(input));
     }
 
-    private JSONObject _getAnswerObject(InputStream istream) throws IOException, JSONException {
-        return _getJSONObject(_getAnswer(istream));
+    private JSONObject _getJSONObject(byte[] array) throws JSONException {
+        return new JSONObject(new JSONTokener(new String(array)));
```
Comment (clement-leprovost-algolia) [inline algoliasearch/src/main/java/com/algolia/search/saas/BaseAPIClient.java]: - `new String(byte[])` without specifying a charset is dangerous. Although the default charset is supposed to be UTF-8 on Android, I would feel more comfortable specifying the charset explicitly.
- `JSONTokener` is superfluous as `JSONObject` can construct itself from a String.

Comment (PLNech) [inline algoliasearch/src/main/java/com/algolia/search/saas/BaseAPIClient.java]: Good points, indeed it won't harm to explicitly ask for UTF-8!

File: algoliasearch/src/main/java/com/algolia/search/saas/BaseIndex.java [hunk 2]
```diff
@@ -342,7 +342,20 @@ protected JSONObject search(Query query) throws AlgoliaException {
...
 
     /**
-     * Wait the publication of a task on the server. 
+     * Search inside the index
+     * @return a byte array containing search results
+     * @throws AlgoliaException
+     */
+    protected byte[] searchRaw(Query query) throws AlgoliaException {
+        String paramsString = query.getQueryString();
+        if (paramsString.length() > 0)
```
Comment (ElPicador) [inline algoliasearch/src/main/java/com/algolia/search/saas/BaseIndex.java]: brackets :wink:

Comment (PLNech) [inline algoliasearch/src/main/java/com/algolia/search/saas/BaseIndex.java]: A code cleanup is planned to cope with these :wink: 

Still, done for this file (Query.java has **55** `if`s without braces !), I'll clean the rest afterwards :smile:

**Comment by PLNech** [comment]
@ElPicador and @clement-leprovost-algolia, does this now LGTY? :smile:

**Comment by ElPicador** [comment]
LGTM from a pure Java perspective

**Comment by clement-leprovost-algolia** [comment]
LTGM

**Pull request merged**